### PR TITLE
[Port] Radiation Rework

### PR DESCRIPTION
**Pull request by XeonMations**
## About The Pull Request

Ports TG Station's:
- [x] https://github.com/tgstation/tgstation/pull/62265
- [x] https://github.com/tgstation/tgstation/pull/62588
- [x] https://github.com/tgstation/tgstation/pull/62550
- [x] https://github.com/tgstation/tgstation/pull/62695
- [x] https://github.com/tgstation/tgstation/pull/63015
- [x] https://github.com/tgstation/tgstation/pull/63152
- [x] https://github.com/tgstation/tgstation/pull/63230
- [x] https://github.com/tgstation/tgstation/pull/64369
- [x] https://github.com/tgstation/tgstation/pull/66305
- [x] https://github.com/tgstation/tgstation/pull/68525
- [x] https://github.com/tgstation/tgstation/pull/61917
- [x] https://github.com/tgstation/tgstation/pull/50530
- [x] https://github.com/tgstation/tgstation/pull/46741
- [x] https://github.com/tgstation/tgstation/pull/49260
- [x] https://github.com/tgstation/tgstation/pull/48935
- [x] https://github.com/tgstation/tgstation/pull/48472
- [x] https://github.com/tgstation/tgstation/pull/52044
- [x] https://github.com/tgstation/tgstation/issues/66639
- [x] https://github.com/tgstation/tgstation/pull/78321

## Why It's Good For The Game

Currently radiation has no proper warnings when you're irradiated, not to mention it takes way too long to get rid of it if you are contaminated. This should fix a few things, while keeping things, hopefully, sane.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Supermatter:

https://github.com/user-attachments/assets/1789180f-41db-430a-9176-3c33d8983f06

Supermatter in proper working condition:

https://github.com/user-attachments/assets/ef313638-223c-4f4a-a39e-1b3641147613

Radiation status effect:

https://github.com/user-attachments/assets/e7c1314e-0215-4be7-a42b-1d0442c3b857

</details>

## Changelog
:cl: Mothblocks, ShizCalev, KittyNoodle, Krysonism, castawaynont, GuillaumePrata, Tiresoup, blessedmulligan, Pickle-Coding, LemonInTheDark, Ghilker, esainane, yoyobatty, vincentiusvin, XeonMations
balance: Radiation is now a status effect, where you are either irradiated or not. When irradiated, you will take constant toxin damage, and irregular burn damage. Radiation is removed once all toxin damage is healed. Standing under a shower will pause the harmful effects of radiation, making it easier to heal. Everything that is irradiated will glow green.
del: Pluoxium can no longer be created by irradiating oxygen and carbon dioxide
del: Removes contamination from radiation.
balance: Uranium structures will irradiate those around them, but will be delayed, giving you time to stay away.
del: The radioactive microlaser no longer irradiates, but still has its knockdown, which was never related.
balance: Screwdriver (the drink) and Potassium Iodide are similar. They will heal toxin damage when the target is irradiated.
balance: Screwdriver, Potassium Iodide, and Pentetic Acid will halt the damaging effects of radiation, just like showers. This means no tox damage or burns.
balance: The gravity generator, instead of giving radiation when on, will now have a high gravity field around it.
balance: Radiation protection has been kept to engineering hardsuits, rad suits.
balance: Genetics no longer uses radiation, but instead genetic damage. Genetic damage works very similarly--once it tops 100%, you will start to take toxin damage. Gorillas are now made through genetic damage instead of radiation.
del: Removed the halogen counter PDA app.
del: Removed the emag effect on geiger counters, though an emag functionality can be added back later.
del: Removed the geiger counter from cyborgs, but mostly because it's not actually really that useful anymore? Everything glows green!
del: Removed the advanced wirebrush, a weird consumer of radiation with an extremely niche purpose. The standard wirebrush still exists.
del: Removed radiation scanning from meson goggles.
del: Removed leftover radiation collectors on ruins.
del: Removed radiation from the radioisotope thermoelectric generator, though it still produces power. Trivia points if you actually know what that is!
balance: The EMP'd advanced energy gun now deals burn damage instead of radiation damage.
balance: Unstable Mutagen now deal toxin damage instead of radiation damage.
balance: Breathing tritium will now irradiate you after a certain amount, as well as deal toxin damage.
del: Removed radiation collectors
fix: Cyborgs and those with rad immunity will no longer get the radiation burn message when near something being dusted by the supermatter.
spellcheck: Rewrote some messages when near something that was dusted by the SM
balance: The further you are from a radiation pulse, and the more objects that block it, the lower your odds are to get irradiated.
balance: Most radiating emitting things have their radiation pulses functionality changed. Being near them is more dangerous, and being far away is less dangerous.
fix: Radiation pulses will ask MC if they should continue processing for every turf they iterate through, rather than just every pulse.
change: The supermatter's power is now harvested in the form of tesla coils.
del: Made the black hole unable to be obtained without adminbus.
qol: supermatter zaps now change color based on the power generated, starting from red and going toward blue the higher the power emitted
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
Well, at least its now possible to integrate actual radiation shielding into modsuits

**Comment by MarkusLarsson421** [comment]
> Currently radiation has no proper warnings when you're irradiated

Isn't that how radiation works? It is known as the silent death.

**Comment by XeonMations** [comment]
And now to fix the SM...

**Comment by Rukofamicom** [comment]
I'm not a fan as a whole, but I'm open to being convinced of the benefits. 

I'd be alright with the removal of the contamination system because it's probably needlessly expensive (I haven't actually checked though) and also contributes greatly to the problem of "invisible" radiation poisoning, but I still much prefer a system where radiation builds up with worsening effects over time as it does. 

I'd be okay with the effects of it building being more apparent and/or the removal of radiation being more simplified, but I don't agree with a more or less total removal of radiation as a mechanic at all. 

:key:

**Comment by PowerfulBacon** [comment]
Radiation is a horrible mechanic and is almost always hated when it actually becomes a threat. I'd be more than happy to see it improved into something that gives the players being abused by it some respect

**Comment by Rukofamicom** [comment]
> I'd be more than happy to see it improved into something that gives the players being abused by it some respect

I'd be happy with an indicator of being radiated too, but it should be more in line with bleeding that stacks, not a binary on/off with no special treatment.

**Comment by PowerfulBacon** [comment]
That would be nice, I was thinking it'd be good to have a counter in the same way as bleeding. It could even cheat like bleeding and have an upper limit without telling the player

File: code/datums/components/irradiated.dm [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+	if (human_parent.getToxLoss() == 0)
+		qdel(src)
+		return PROCESS_KILL
+
+	if (should_halt_effects(parent))
+		return
+
+	if (human_parent.stat > DEAD)
+		human_parent.dna?.species?.handle_radiation(human_parent, world.time - beginning_of_irradiation, delta_time)
```
Comment (PowerfulBacon) [inline code/datums/components/irradiated.dm:98]: Having a stat greater than dead is impossible

File: code/__DEFINES/traits/declarations.dm [hunk 2]
```diff
@@ -309,4 +308,28 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
...
+#define TRAIT_IRRADIATED "iraddiated"
+
+/// Immune to being irradiated
+#define TRAIT_RADIMMUNE "rad_immunity"
+
+/// Harmful radiation effects, the toxin damage and the burns, will not occur while this trait is active
+#define TRAIT_HALT_RADIATION_EFFECTS "halt_radiation_effects"
+
+//If the species gains good things from radiation. See: Dionae
+#define TRAIT_RADBOOSTER "rad_booster"
```
Comment (PowerfulBacon) [inline code/__DEFINES/traits/declarations.dm:330]: The name of this trait doesn't really make much sense. I also think that theres a bit of an issue where if you have a species that heals from rads and forget to put the trait that stops you from getting the irradiated component then your species will get permanent immunity which is a bit of a trap for future coders but I guess its fine

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
## Supermatter Power Output

The power output from the SM is ridiculous, and the radiation it makes is huge.
With a roundstart N2 setup:

image

And radiation filled the entire room until this point

image

At the very least, the power output shouldn't be intended since the SM currently makes in the range of 100-300kW

## Radiation Effects

I dislike how clearing radiation works. Here is my feedback:
- Make radiation like the bleeding status effect, where it has an amount attached to it (give it a limit)
- Tick the counter up as you increase radiation instead of using a prob() that randomly gives the effect
- If the effect is less than a threshold, don't throw the alert and do nothing (at this point it should slowly decrease over time until it goes away)
- One you are at a certain amount, then start doing the toxin damage
- Showering lowers that value, all the rad-meds lower the value of radiation. (Maybe either no toxin damage should increase the rate at which rads go away, or washing/changing clothes should somehow have an effect, but the clothes one is harder and may be out of reasonable scope).
- Remove the halting of the effect entirely
- If you get a lot of rads and start approaching the limit, then start doing some of the other radiation fluff like nosebleeds and hair clumps falling out (these don't have to actually do anything, but are there for flavour)

File: code/__DEFINES/traits/declarations.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/__DEFINES/traits/declarations.dm:330]: I see what the intention is for this now, TRAIT_ALWAYS_PROCESS_RADS, TRAIT_BENEFICIAL_RADS or something might make more sense

**Comment by XeonMations** [comment]
Making this a draft once more, will need to work on this a lot after LINDA is done.

**Comment by XeonMations** [comment]
was told to close this until i decide to work on it once more

**Comment by Rukofamicom** [comment]
Took away the stale label so you should be able to freely re-open it without issues when you do get to it

**Pull request closed without merging**